processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -604,30 +608,43 @@ else if ( !sourceReference.isNested() ) {
             }
         }
 
-        private String getSourcePresenceCheckerRef( SourceReference sourceReference ) {
-            String sourcePresenceChecker = null;
+        private PresenceCheck getSourcePresenceCheckerRef(SourceReference sourceReference ) {
+            PresenceCheck sourcePresenceChecker = null;
             if ( !sourceReference.getPropertyEntries().isEmpty() ) {
                 Parameter sourceParam = sourceReference.getParameter();
                 // TODO is first correct here?? shouldn't it be last since the remainer is checked
                 // in the forged method?
                 PropertyEntry propertyEntry = sourceReference.getShallowestProperty();
                 if ( propertyEntry.getPresenceChecker() != null ) {
-                    sourcePresenceChecker = sourceParam.getName()
-                        + "." + propertyEntry.getPresenceChecker().getSimpleName() + "()";
+                    List<PresenceCheck> presenceChecks = new ArrayList<>();
+                    presenceChecks.add( new SourceReferenceMethodPresenceCheck(
+                        sourceParam.getName(),
+                        propertyEntry.getPresenceChecker().getSimpleName()
+                    ) );
 
                     String variableName = sourceParam.getName() + "."
                         + propertyEntry.getReadAccessor().getSimpleName() + "()";
                     for (int i = 1; i < sourceReference.getPropertyEntries().size(); i++) {
                         PropertyEntry entry = sourceReference.getPropertyEntries().get( i );
                         if (entry.getPresenceChecker() != null && entry.getReadAccessor() != null) {
-                            sourcePresenceChecker += " && " + variableName + " != null && "
-                                + variableName + "." + entry.getPresenceChecker().getSimpleName() + "()";
+                            presenceChecks.add( new NullPresenceCheck( variableName ) );
```

Comment:
I see that we add a `NullPresenceCheck` and a `SourceReferenceMethodPresenceCheck` .. Why not assume delegation of the null check to a default `SourceReferenceMethodPresenceCheck`? Wouldn't that give
1. More room to play for the implementor of `SourceReferenceMethodPresenceCheck`? (null is a value, right)
2. Give more readable code?

Or perhaps I don't understand it sufficient yet at this point in time (first review pass).

Reply:
This is a special one which is only did for #1456. My idea was to make the presence checks independent of each other. The `SourceReferenceMethodPresenceCheck` is also used in the case when we need to call `hasProperty()` on a source bean. In this case we shouldn't do a `null` check for the `source` since it has already been done at the beginning of the method. We want to avoid the following scenario


```
public Target map(Source source) {

    if ( source == null ) {
        return null;
    }
    
    Target target = new Target();
    
    if ( source != null && source.hasName() ) {
        target.setName( source.getName() );
    }

}
```

Reply:
> The `SourceReferenceMethodPresenceCheck` is also used in the case when we need to call `hasProperty()` on a source bean

So, the `hasXYZ` is one of the 'default' presence checkers? Is this overridden as well as you define a `conditionBy`? (so if you've defined `has` methods.. what will happen in relation to the new stuff)?

Reply:
Yes `hasXYZ` is one of the 'default' presence checkers. If you define a custom `@Condition` method that is applicable for that type then, this will be used, i.e. your presence checker on the bean won't be used. Perhaps, those should have highest precedence instead?